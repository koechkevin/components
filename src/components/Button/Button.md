# Button

/src/components/Button/Button.tsx

``` TS
<Button>This is a Button</Button>
```
