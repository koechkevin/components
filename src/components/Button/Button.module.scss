@import '../../styles/global';

.button {
  &:global([ant-click-animating-without-extra-node]:after) {
    animation: 0s;
  }

  @mixin commonDimensions {
    font-size: 16px;
    line-height: 24px;
    border-radius: 8px;
    font-weight: 600;
    padding: 0 16px;
  }

  &:global(.ant-btn) {
    height: 40px;
    border: none;
    box-shadow: none;
    color: $color-base-dark-gray;
    @include commonDimensions();
    background-color: $color-basic-light-gray;

    &[disabled],
    &[disabled]:hover {
      color: rgba($color: $color-base-dark-gray, $alpha: 0.38);
    }
  }

  &:global(.ant-btn-background-ghost) {
    border: solid 1px $color-primary;
  }

  &:global(.ant-btn-primary) {
    border-color: $color-primary;
    background-color: $color-primary;
    color: $color-white;
    @include commonDimensions();

    &:hover,
    &:active,
    &:focus {
      border-color: $color-primary-dark;
      background-color: $color-primary-dark;
    }

    &[disabled],
    &[disabled]:hover {
      opacity: 0.38;
      color: $color-white;
      border-color: $color-primary;
      background-color: $color-primary;
    }
  }

  &:global(.ant-btn-background-ghost) {
    @include commonDimensions();
    color: $color-primary;
    border-color: $color-primary;

    &:hover,
    &:active,
    &:focus {
      color: $color-primary-dark;
      border-color: $color-primary-dark;
    }

    &[disabled],
    &[disabled]:hover {
      opacity: 0.38;
      color: $color-primary;
      border-color: $color-primary;
    }
  }

  &:global(.ant-btn-danger) {
    @include commonDimensions();
    color: $color-white;
    border-color: $color-error-secondary;
    background-color: $color-error-secondary;

    &:hover,
    &:active,
    &:focus {
      border-color: $color-error-hover-secondary;
      background-color: $color-error-hover-secondary;
    }

    &[disabled],
    &[disabled]:hover {
      opacity: 0.38;
      color: $color-white;
      border-color: $color-error-secondary;
      background-color: $color-error-secondary;
    }
  }

  &:global(.ant-btn-danger.ant-btn-background-ghost) {
    color: $color-error-secondary;
    border-color: $color-error-secondary;

    &:hover,
    &:active,
    &:focus {
      border-color: $color-error-hover-secondary;
      background-color: $color-error-hover-secondary;
    }

    &[disabled],
    &[disabled]:hover {
      opacity: 0.38;
      color: $color-white;
      border-color: $color-error-secondary;
      background-color: $color-error-secondary;
    }
  }

  &:global(.ant-btn-round) {
    font-size: 16px;
    font-weight: bold;
    padding: 0 15px;
    border-radius: 40px;
    border-color: $color-primary;
    background-color: $color-primary;

    &:hover,
    &:active,
    &:focus {
      border-color: $color-primary-dark;
      background-color: $color-primary-dark;
    }

    &:global(.ant-btn-sm) {
      border-radius: 40px;
    }
  }

  &:global(.ant-btn-link) {
    font-weight: bold;
    color: rgba($color: $color-base-dark, $alpha: 0.6);
    background-color: unset;

    &:hover,
    &:active,
    &:focus {
      color: $color-primary;
      border-color: $color-primary;
    }
  }

  &:global(.ant-btn-icon-only) {
    border: none;
    color: $color-primary;
    font-weight: bold;
    box-shadow: none;

    &:hover,
    &:active,
    &:focus {
      background-color: $color-primary-light;
    }

    &[disabled],
    &[disabled]:hover {
      opacity: 0.38;
      color: $color-primary;
      background-color: transparent;
    }
  }

  &:global(.ant-btn-round.ant-btn-lg),
  &:global(.ant-btn-lg) {
    height: 44px;
  }

  &:global(.ant-btn-sm) {
    height: 32px;
    font-size: 14px;
    border-radius: 4px;
    padding: 0 12px;
  }

  &:global(.ant-btn-loading) {
    &::before {
      opacity: 0;
    }
  }
}

.loading {
  &:global(.ant-btn) {
    padding: 0 24px;
  }

  pointer-events: none;
}

:global .fa-spinner-third {
  vertical-align: middle;
}
