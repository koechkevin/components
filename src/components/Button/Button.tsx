import { faSpinnerThird } from '@fortawesome/pro-duotone-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Button as AntdButton } from 'antd';
import { ButtonProps } from 'antd/lib/button';
import React, { FC } from 'react';

import styles from './Button.module.scss';

const Button: FC<ButtonProps> = (props) => {
  const { loading, children, className, ...resetProps } = props;
  const classNames = [styles.button, loading ? styles.loading : '', className].join(' ');

  return (
    <AntdButton className={classNames} {...resetProps}>
      {loading ? (
        <FontAwesomeIcon style={{ fontSize: 22, animationDuration: '1.2s' }} icon={faSpinnerThird} spin />
      ) : (
        children
      )}
    </AntdButton>
  );
};

export default Button;
