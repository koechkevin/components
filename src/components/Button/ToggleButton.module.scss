@import '../../styles/global';

.toggle {
  cursor: pointer;
  outline: none;
  border: none;
  width: 36px;
  height: 36px;
  border-radius: 4px;
  font-size: 20px;
}

.toggleOn {
  color: white;
  background: $color-primary;

  &:hover {
    background-color: $color-primary-dark;
  }
}

.toggleOff {
  background-color: transparent;
  color: $color-primary;

  &:hover {
    background-color: $color-primary-light;
  }
}

.disabled {
  color: #a8c4ed !important;
  pointer-events: none !important;
  cursor: none;
}