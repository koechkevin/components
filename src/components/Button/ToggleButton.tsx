import React, { CSSProperties, FC, useState } from 'react';

import styles from './ToggleButton.module.scss';

interface Props {
  style?: CSSProperties;
  checked?: boolean;
  className?: string;
  disabled?: boolean;
  onClick?: () => void;
}

export const ToggleButton: FC<Props> = (props) => {
  const { checked, className, disabled, onClick, style } = props;
  const [isChecked, setIsChecked] = useState(checked);
  const classNames = [
    styles.toggle,
    isChecked ? styles.toggleOn : styles.toggleOff,
    className,
    disabled ? styles.disabled : '',
  ].join(' ');
  const click = (e: any) => {
    e.preventDefault();
    setIsChecked(!isChecked);
    onClick && onClick();
  };

  return (
    <button onClick={click} className={classNames} style={style}>
      &times;
    </button>
  );
};

export default ToggleButton;
