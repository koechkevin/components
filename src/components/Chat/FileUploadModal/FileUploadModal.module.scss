@import '../../../styles/global';

.fileContainer {
  background: $color-gray-lighter;
  border-radius: 5px;
  padding: 8px;

  .content {
    height: 60px;
    padding: 16px;
    flex-wrap: nowrap;
    overflow: hidden;
    border: 1px solid $color-primary-light;
    background-color: $color-white;
  }
}

.textArea {
  margin-top: 16px;

  :global {
    textarea {
      resize: none;
      min-height: 80px;
      color: rgba($color: $color-base-dark, $alpha: 1);
    }

    .ant-input:focus {
      border-color: $color-primary;
      outline: 0;
      -webkit-box-shadow: none;
      box-shadow: none;
    }

    .ant-input:hover {
      border-color: $color-primary;
    }
  }
}

.file {
  .icon {
    color: $color-primary;
  }
}

.fileName {
  @include ellipsis;
  padding-left: 8px;
}
