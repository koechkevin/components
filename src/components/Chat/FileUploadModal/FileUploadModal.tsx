import { Col, Row } from 'antd';
import { ModalProps } from 'antd/lib/modal';
import { RcFile } from 'antd/lib/upload';
import React, { ChangeEvent, FC, MouseEvent, useEffect, useState } from 'react';
import TextArea from 'react-textarea-autosize';
import { useMedia } from 'react-use';

import { FileExtension, selectIcon } from '../../../utils';
import { Icon } from '../../Icon';
import { Modal } from '../../Modal';
import { Tooltip } from '../../Tooltip';

import styles from './FileUploadModal.module.scss';

interface Props extends ModalProps {
  file: RcFile;
  text?: string;
  extension?: FileExtension;
}

const FileUploadModal: FC<Props> = (props) => {
  const { extension, text, file, onOk, onCancel, ...otherProps } = props;
  const [value, setValue] = useState<any>(text);
  const isMobile = useMedia('(max-width: 575px)');

  useEffect(() => {
    if (text) {
      setValue(text);
    }
  }, [setValue, text]);

  const handleCancel = (e: MouseEvent<HTMLElement>) => {
    onCancel && onCancel(e);
  };

  const handleOk = () => {
    onOk && onOk(value);
  };

  return (
    <Modal
      centered
      width={500}
      okText="Upload"
      onOk={handleOk}
      title="Upload a file"
      onCancel={handleCancel}
      afterClose={() => setValue('')}
      {...otherProps}
    >
      <Row>
        <Row className={styles.fileContainer}>
          <Row className={styles.content} type="flex" align="middle">
            <Col className={styles.file}>
              <Icon icon={selectIcon(extension)} className={styles.icon} />
            </Col>
            <Tooltip title={file?.name}>
              <Col className={styles.fileName}>{file?.name}</Col>
            </Tooltip>
          </Row>
        </Row>
        <Row className={styles.textArea}>
          <TextArea
            min={3}
            autoFocus
            value={value}
            className="ant-input"
            maxRows={isMobile ? 5 : 8}
            style={{ minHeight: 80 }}
            onChange={(e: ChangeEvent<HTMLTextAreaElement>) => setValue(e.target.value)}
          />
        </Row>
      </Row>
    </Modal>
  );
};

export default FileUploadModal;
