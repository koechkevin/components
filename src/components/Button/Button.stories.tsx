import { Card, Col, Row, Typography } from 'antd';
import React, { CSSProperties } from 'react';

import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';

import Button from './Button';
import notes from './Button.md';
import ToggleButton from './ToggleButton';

const { Title, Text } = Typography;
const { Grid } = Card;
const gridStyle: CSSProperties = {
  width: '25%',
  textAlign: 'center',
};
const gridStyle1: CSSProperties = {
  width: '33.333333%',
  textAlign: 'center',
};
const gridStyle2: CSSProperties = {
  width: '33.333333%',
  textAlign: 'center',
  backgroundColor: '#0050c8',
};

const toggleStyleNegative: CSSProperties = {
  float: 'right',
  fontSize: '12px',
  color: 'white',
  top: '100%',
  marginTop: '20%',
};

storiesOf('Button', module)
  .addDecorator((story) => <div style={{ padding: 24 }}>{story()}</div>)
  .add(
    'primary',
    () => (
      <>
        <Title level={2}>Button states</Title>
        <Row gutter={24}>
          <Col>
            <Card bordered={false} title="Primary Button">
              <Grid style={gridStyle}>
                <Button type="primary" onClick={action('onClick')}>
                  Enabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary">Hover</Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" disabled>
                  Disabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" loading>
                  Loading
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" size="large">
                  Enabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" size="large">
                  Hover
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" size="large" disabled>
                  Disabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" size="large" loading>
                  Loading
                </Button>
              </Grid>
            </Card>
          </Col>
          <Col style={{ marginTop: 24 }}>
            <Card bordered={false} title="Negative Button">
              <Grid style={gridStyle}>
                <Button type="primary" size="small" ghost>
                  Enabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" size="small" ghost>
                  Hover
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" size="small" ghost disabled>
                  Disabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" size="small" ghost loading>
                  Loading
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button size="large" type="primary" ghost>
                  Enabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button size="large" type="primary" ghost>
                  Hover
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button size="large" type="primary" ghost disabled>
                  Disabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button size="large" type="primary" ghost loading>
                  Loading
                </Button>
              </Grid>
            </Card>
          </Col>
          <Col style={{ marginTop: 24 }}>
            <Card bordered={false} title="Danger Button">
              <Grid style={gridStyle}>
                <Button type="danger">Enabled</Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="danger">Hover</Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="danger" disabled>
                  Disabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="danger" loading>
                  Loading
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button size="large" type="danger">
                  Enabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button size="large" type="danger">
                  Hover
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button size="large" type="danger" disabled>
                  Disabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button size="large" type="danger" loading>
                  Loading
                </Button>
              </Grid>
            </Card>
          </Col>
          <Col style={{ marginTop: 24 }}>
            <Card bordered={false} title="Default Buttons">
              <Grid style={gridStyle}>
                <Button size="small">Enabled</Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button size="small">Hover</Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button size="small" disabled>
                  Disabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button size="small" loading>
                  Loading
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button>Enabled</Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button>Hover</Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button disabled>Disabled</Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button loading>Loading</Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button size="large">Enabled</Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button size="large">Hover</Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button size="large" disabled>
                  Disabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button size="large" loading>
                  Loading
                </Button>
              </Grid>
            </Card>
          </Col>
          <Col style={{ marginTop: 24 }}>
            <Card bordered={false} title="Round Button">
              <Grid style={gridStyle}>
                <Button type="primary" size="small" shape="round">
                  Enabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" size="small" shape="round">
                  Hover
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" size="small" shape="round" disabled>
                  Disabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" size="small" shape="round" loading>
                  Loading
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" shape="round" onClick={action('onClick')}>
                  Enabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" shape="round">
                  Hover
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" shape="round" disabled>
                  Disabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" shape="round" loading>
                  Loading
                </Button>
              </Grid>
            </Card>
          </Col>
          <Col style={{ marginTop: 24 }}>
            <Card bordered={false} title="Link Button">
              <Grid style={gridStyle}>
                <Button type="link">Enabled</Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="link">Hover</Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="link" disabled>
                  Disabled
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="link" loading>
                  Loading
                </Button>
              </Grid>
            </Card>
          </Col>
          <Col style={{ marginTop: 24 }}>
            <Card bordered={false} title="Toggle Button">
              <Grid style={gridStyle1}>
                <ToggleButton checked={false} />
                <Text style={{ ...toggleStyleNegative, color: 'black' }}>Enabled</Text>
              </Grid>
              <Grid style={gridStyle1}>
                <ToggleButton style={{ backgroundColor: 'rgb(223, 233, 248)' }} checked={false} />
                <Text style={{ ...toggleStyleNegative, color: 'black' }}>Hover</Text>
              </Grid>
              <Grid style={gridStyle1}>
                <ToggleButton disabled checked={false} />
                <Text style={{ ...toggleStyleNegative, color: 'black' }}>Disabled</Text>
              </Grid>
              <Grid style={gridStyle2}>
                <ToggleButton onClick={action('onClick')} checked />
                <Text style={toggleStyleNegative}>Enabled</Text>
              </Grid>
              <Grid style={gridStyle2}>
                <ToggleButton style={{ backgroundColor: 'rgb(0, 61, 150)' }} checked />
                <Text style={toggleStyleNegative}>Hover</Text>
              </Grid>
              <Grid style={gridStyle2}>
                <ToggleButton checked disabled />
                <Text style={toggleStyleNegative}>Disabled</Text>
              </Grid>
            </Card>
          </Col>
          <Col style={{ marginTop: 24 }}>
            <Card bordered={false} title="Mobile Button">
              <Grid style={gridStyle}>
                <Button type="primary" block style={{ marginBottom: 24 }}>
                  Enabled
                </Button>
                <Button type="primary" block style={{ marginBottom: 24 }}>
                  Hover
                </Button>
                <Button type="primary" disabled block style={{ marginBottom: 24 }}>
                  Disabled
                </Button>
                <Button type="primary" loading block style={{ marginBottom: 24 }}>
                  Loading
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="primary" ghost block style={{ marginBottom: 24 }}>
                  Enabled
                </Button>
                <Button type="primary" ghost block style={{ marginBottom: 24 }}>
                  Hover
                </Button>
                <Button type="primary" ghost disabled block style={{ marginBottom: 24 }}>
                  Disabled
                </Button>
                <Button type="primary" ghost loading block style={{ marginBottom: 24 }}>
                  Loading
                </Button>
              </Grid>
              <Grid style={gridStyle}>
                <Button type="danger" block style={{ marginBottom: 24 }}>
                  Enabled
                </Button>
                <Button type="danger" block style={{ marginBottom: 24 }}>
                  Hover
                </Button>
                <Button type="danger" disabled block style={{ marginBottom: 24 }}>
                  Disabled
                </Button>
                <Button type="danger" loading block style={{ marginBottom: 24 }}>
                  Loading
                </Button>
              </Grid>
            </Card>
          </Col>
        </Row>
      </>
    ),
    {
      notes: { markDown: notes },
    },
  );
