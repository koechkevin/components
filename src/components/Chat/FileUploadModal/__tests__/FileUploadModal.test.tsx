import { RcFile } from 'antd/lib/upload';
import { mount } from 'enzyme';
import React from 'react';

import { FileExtension } from '../../../../utils';
import { Button } from '../../../Button';
import FileUploadModal from '../FileUploadModal';

export const file = {
  uid: '',
  size: 1024,
  name: 'Jane_Doe_Cv.pdf',
  lastModified: 1234,
  type: 'application/pdf',
  webkitRelativePath: '',
  lastModifiedDate: new Date(),
};

describe('Tests for file upload modal', () => {
  const props = {
    title: 'modal title',
    name: 'Jane_Doe_CV.pdf',
    onCancel: jest.fn(),
    onOk: jest.fn(),
    visible: true,
    file: file as RcFile,
    extension: FileExtension.pdf,
  };

  const wrapper = mount(<FileUploadModal {...props} />);

  it('should render the file upload modal without crushing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should simulate click of cancel button', () => {
    const cancelButton = wrapper.find(Button).first();
    cancelButton.simulate('click');

    expect(cancelButton.contains('Cancel')).toEqual(true);
    expect(props.onCancel).toHaveBeenCalled();
  });

  it('should simulate clicking of upload button', () => {
    const okButton = wrapper.find(Button).last();
    okButton.simulate('click');

    expect(okButton.contains('Upload')).toEqual(true);
    expect(props.onOk).toHaveBeenCalled();
  });
});
