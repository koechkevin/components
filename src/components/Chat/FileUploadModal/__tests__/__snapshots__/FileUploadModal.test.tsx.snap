// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for file upload modal should render the file upload modal without crushing 1`] = `ReactWrapper {}`;
